### ISSUE TITLE
Add associated type to link request and response types?

### ISSUE THREAD

**Issue by icefoxen** [ferristseng/rust-ipfs-api#6]
Just a thought that I had while fixing a bug I'd made...

We have the `request::ApiRequest` type which specifies some data for the request type.  It should be possible to make an associated type `request::ApiRequest::Response` to link the response type to a particular request type.  Then the type of `client::IpfsClient::request<Req, Res>() -> AsyncResponse<Res>` could just become `client::IpfsClient::request<Req>() -> AsyncResponse<Req::Response>`.

That wouldn't have prevented the bug that I'd made, so I'm not sure if this will actually help rather than just bikeshedding, but might make life a little simpler?

...actually, now that I think about it, we might be able to make the `ApiRequest::path()` method be an associated const now.  Though that might be more restrictive than we want?  I dunno.

**Comment by ferristseng** [comment]
Yeah, I like the idea of the associated type. I was planning on making a macro to implement the `ApiRequest` trait, since a large portion of it is just boiler plate. Since I was pretty consistent with the type naming, I think it might be possible to make a macro where you just pass in the request name, and it would automatically link the response as an associated type. I think a combination of the associated type, and the macro would make it really difficult to ever make a naming issue.

Also I'm okay with making path an associated const. I didn't know about associated consts, but the path method is pretty much acting like one right now.

**Comment by icefoxen** [comment]
> I think it might be possible to make a macro where you just pass in the request name, and it would automatically link the response as an associated type

Let me know if you work this out, but afaik synthesizing new identifiers in macros is not currently feasible...  :-(
